File: apps/web/src/lib/portone.ts
/**
 * PortOne V2 REST API 래퍼
 * @see https://developers.portone.io/api/rest-v2
 */

const PORTONE_API_BASE = 'https://api.portone.io';

function getHeaders() {
  const secret = process.env.PORTONE_V2_SECRET_KEY;
  if (!secret) throw new Error('PORTONE_V2_SECRET_KEY is not set');
  return {
    'Content-Type': 'application/json',
    Authorization: `PortOne ${secret}`,
  };
}

// 결제 정보 조회
export type PortOnePayment = {
  id: string;
  status: string;
  amount: { total: number; paid: number };
  method?: { type: string; billingKey?: string };
  customer?: { id: string; email: string };
  customData?: string;
};

export async function getPayment(paymentId: string): Promise<PortOnePayment> {
  const res = await fetch(`${PORTONE_API_BASE}/payments/${encodeURIComponent(paymentId)}`, {
    method: 'GET',
    headers: getHeaders(),
  });
  if (!res.ok) {
    const body = await res.text();
    throw new Error(`PortOne getPayment failed: ${res.status} ${body}`);
  }
  return res.json();
}

// 빌링키 결제 (정기결제 갱신)
export async function payWithBillingKey(params: {
  paymentId: string;
  billingKey: string;
  orderName: string;
  amount: number;
  currency?: string;
  customerId: string;
}): Promise<PortOnePayment> {
  const res = await fetch(`${PORTONE_API_BASE}/payments/${encodeURIComponent(params.paymentId)}/billing-key`, {
    method: 'POST',
    headers: getHeaders(),
    body: JSON.stringify({
      billingKey: params.billingKey,
      orderName: params.orderName,
      amount: { total: params.amount },
      currency: params.currency ?? 'KRW',
      customer: { id: params.customerId },
    }),
  });
  if (!res.ok) {
    const body = await res.text();
    throw new Error(`PortOne billingKey payment failed: ${res.status} ${body}`);
  }
  return res.json();
}

// 결제 취소 (환불)
export async function cancelPayment(params: {
  paymentId: string;
  reason: string;
  amount?: number;
}): Promise<{ cancellation: { id: string; totalAmount: number } }> {
  const res = await fetch(`${PORTONE_API_BASE}/payments/${encodeURIComponent(params.paymentId)}/cancel`, {
    method: 'POST',
    headers: getHeaders(),
    body: JSON.stringify({
      reason: params.reason,
      amount: params.amount,
    }),
  });
  if (!res.ok) {
    const body = await res.text();
    throw new Error(`PortOne cancelPayment failed: ${res.status} ${body}`);
  }
  return res.json();
}

// 웹훅 시그니처 검증
export function verifyWebhookSignature(
  body: string,
  signature: string | null,
  secret: string
): boolean {
  if (!signature) return false;
  // PortOne V2 웹훅은 HMAC-SHA256 시그니처 사용
  // 실제 환경에서는 crypto.subtle 또는 node:crypto 사용
  try {
    const crypto = require('node:crypto');
    const expected = crypto
      .createHmac('sha256', secret)
      .update(body)
      .digest('hex');
    return crypto.timingSafeEqual(
      Buffer.from(signature),
      Buffer.from(expected)
    );
  } catch {
    return false;
  }
}

File: apps/api/app/services/portone.py
"""PortOne V2 결제 검증 클라이언트."""

from typing import Any, TypedDict, cast

import httpx

from app.config import settings

PORTONE_API_BASE = "https://api.portone.io"


class PaymentVerification(TypedDict):
    status: str
    amount: int
    currency: str
    method: str
    paid_at: str | None


async def verify_payment(payment_id: str) -> PaymentVerification:
    """PortOne V2 API로 결제 상태 검증.

    Returns:
        dict with keys: status, amount, currency, method, paid_at
    Raises:
        ValueError if payment not found or not paid
    """
    async with httpx.AsyncClient() as client:
        resp = await client.get(
            f"{PORTONE_API_BASE}/payments/{payment_id}",
            headers={
                "Authorization": f"PortOne {settings.PORTONE_API_SECRET}",
                "Content-Type": "application/json",
            },
        )

        if resp.status_code == 404:
            raise ValueError("결제 정보를 찾을 수 없습니다")

        resp.raise_for_status()
        data = cast(dict[str, Any], resp.json())

        status = data.get("status")
        if status != "PAID":
            raise ValueError(f"결제가 완료되지 않았습니다 (status: {status})")

        amount = cast(dict[str, Any], data.get("amount", {}))
        method = cast(dict[str, Any], data.get("method", {}))

        return {
            "status": status,
            "amount": cast(int, amount.get("total", 0)),
            "currency": cast(str, data.get("currency", "KRW")),
            "method": cast(str, method.get("type", "")),
            "paid_at": cast(str | None, data.get("paidAt")),
        }


async def verify_payment_amount(payment_id: str, expected_amount: int) -> PaymentVerification:
    """결제 검증 + 금액 일치 확인."""
    payment = await verify_payment(payment_id)
    if payment["amount"] != expected_amount:
        raise ValueError(f"결제 금액 불일치: 예상 {expected_amount}원, 실제 {payment['amount']}원")
    return payment
